src/unistd.rs

```diff
@@ -2136,3 +2146,424 @@ mod setres {
+                gid: Gid::from_raw((*gr).gr_gid),
+                mem: Group::members((*gr).gr_mem)
+            }
+        }
+    }
+}
+
+impl Group {
+    unsafe fn members(mem: *mut *mut c_char) -> Vec<String> {
+        let mut ret = vec![];
+
+        for i in 0.. {
+            let u = mem.offset(i);
+            if (*u).is_null() {
+                break;
+            } else {
+                let s = CStr::from_ptr(*u).to_string_lossy().into_owned();
+                ret.push(s);
+            }
+        }
+
+        ret
+    }
+}
+
+#[derive(Debug)]
+/// Query a user.
+///
+/// The buffer size variants should not be needed 99% of the time; the default buffer size of 1024
+/// should be more than adequate. Only use the buffer size variants if you receive an ERANGE error
+/// without them, or you know that you will be likely to.
+///
+/// # Examples
+///
+/// ```
+/// use nix::unistd::{User, UserQuery, Queryable};
+/// let find = String::from("root");
+/// // Returns an Option<Result<User>>, thus the double unwrap.
+/// let res = User::query( UserQuery::Name(find.clone()) ).unwrap().unwrap();
+/// assert!(res.name == find);
+/// ```
+pub enum UserQuery {
+    /// Get a user by UID. 
+    ///
+    /// Internally, this function calls
+    /// [getpwuid_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Uid(Uid),
+    /// Get a user by name. 
+    ///
+    /// Internally, this function calls
+    /// [getpwnam_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Name(String),
+    UidWithBufsize(Uid, usize),
+    NameWithBufsize(String, usize)
+}
+
+/// Query a group. 
+///
+/// The buffer size variants should not be needed 99% of the time; the default buffer size of 1024
+/// should be more than adequate. Only use the buffer size variants if you receive an ERANGE error
+/// without them, or you know that you will be likely to.
+///
+/// # Examples
+///
+/// ```ignore
+/// use nix::unistd::{Group, GroupQuery};
+/// // On many systems there's no `root` group; on FreeBSD for example it's known as `wheel`
+/// let find = String::from("root");
+/// // Returns an Option<Result<Group>>, thus the double unwrap. Will panic if there's no `root`
+/// // group.
+/// let res = Group::query( GroupQuery::Name(find.clone()) ).unwrap().unwrap();
+/// assert!(res.name == find);
+/// ```
+#[derive(Debug)]
+pub enum GroupQuery {
+    /// Get a group by GID. 
+    ///
+    /// Internally, this function calls
+    /// [getgrgid_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Gid(Gid),
+    /// Get a group by name. 
+    ///
+    /// Internally, this function calls
+    /// [getgrnam_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Name(String),
+    GidWithBufsize(Gid, usize),
+    NameWithBufsize(String, usize)
+}
+
+pub trait Queryable<Q> where Self: ::std::marker::Sized {
+    fn query(q: Q) -> Option<Result<Self>>;
```

Comment:
This probably should be public, yes?

Additionally, what's with the weird `Option<Result<_>>` type?

Reply:
The `Option<Result<_>>` type makes sense in this case due to the lower level API.

If you receive `Some(Ok(_))`, it means that *there was a user found* and *its data fit in the buffer and is now wrapped in the result*.
If you receive a `Some(Err(_))`, it means that *there was a user found* but *its data did not fit in the buffer*, and you should recall with one of the `WithBufsize` variants with a higher `size`.
If you receive a `None`, it means that no user was found.

Reply:
I think a better API would be for the caller to supply an optional bufsize.  If not provided, then Nix will choose a bufsize, retrying with a larger buffer if necessary.  Also, why is there a need for the `Queryable` trait?  I get that `User` and `Group` have a similar API.  But if there's never going to be any generic function with an indiscriminate `Queryable` type in its signature, then there's no need to define a trait.  I might do it like this:
```rust
pub enum GroupQuery<'a> {
    Gid(gid),
    Name(&'a str)
}
impl Group {
    fn query(q: GroupQuery, bufsize: Option<usize>) -> Option<Result<Self>>;
}

let res = Group::query(GroupQuery::Name("root"), Some(2048)).unwrap();
```
However, it would be even simpler to eliminate the `GroupQuery` type entirely, replacing it with two separate constructors:
```rust
impl Group {
    fn from_gid(gid: Gid, bufsize: Option<usize>) -> Option<Result<Self>>;
    fn from_name(name: &str, bufsize: Option<usize>) -> Option<Result<Self>>;
}

let res = Group::from_name("root", Some(2048)).unwrap();
```
What @Susurrus was suggesting would be to create a bespoke return type to get rid of the double unwrap wart.  It would look like this:
```
enum GroupResult {
    /// No group found
    None,
    /// An error was encountered while searching
    Err(Error),
    Some(Group)
}
```
Such a type could still support most of the common operators that are defined for `std::option::Option` and `std::result::Result`, i.e. `unwrap`, `is_some`, `is_err`, `is_none`, etc.